seaworthy/client.py

```diff
@@ -0,0 +1,88 @@
+import hyperlink
+import requests
+
+
+class ContainerClient:
+    URL_DEFAULTS = {'scheme': 'http'}
+
+    def __init__(self, host, port, url_defaults=None, session_factory=None):
+        """
+        :param host:
+            The address for the host to connect to.
+        :param port:
+            The port for the host to connect to.
+        :param dict url_defaults:
+            Parameters to default to in the generated URLs, see
+            `~hyperlink.URL`.
+        :param session_factory:
+            A no-args callable that returns Requests' Session objects.
+        """
+        if session_factory is None:
+            session_factory = requests.Session
+        self._session_factory = session_factory
+
+        _url_defaults = self.URL_DEFAULTS.copy()
+        if url_defaults is not None:
+            _url_defaults.update(url_defaults)
+        self._base_url = hyperlink.URL(
+            host=host, port=int(port), **_url_defaults)
+
+        self._session = None
+
+    def __enter__(self):
+        self._get_session()
+        return self
+
+    def close(self):
+        """
+        Closes the underlying Session object.
+        """
+        if self._session is not None:
+            self._session.close()
+            self._session = None
+
+    def __exit__(self, *args):
+        self.close()
+
+    def _get_session(self):
+        if self._session is None:
+            self._session = self._session_factory()
```

Comment:
I wasn't entirely sure where to create these sessions... With requests, you can still make requests with a session object that has been closed (I assume it is reopened or something?).

Reply:
I think just create the session in `__init__()` and close it in `close()` (which is called from `__exit__()`. There doesn't seem to be any real cost to having an unused session lying around, and all closing the session does is close all the lurking connections and such.